f3/certstore.go

```diff
@@ -0,0 +1,189 @@
+package f3
+
+import (
+	"context"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/ipfs/go-datastore"
+	"github.com/ipfs/go-datastore/namespace"
+
+	"github.com/Kubuxu/go-broadcast"
+	"golang.org/x/xerrors"
+)
+
+// TODO: setup cbor-gen
+type Cert struct {
+	Instance uint64
+}
+
+func (c *Cert) UnmarshalBinary(b []byte) error {
+	// temp until cbor-gen
+	if len(b) != 8 {
+		return xerrors.Errorf("expected length 8 got %d", len(b))
+	}
+	c.Instance = binary.LittleEndian.Uint64(b)
+	return nil
+}
+func (c Cert) MarshalBinary() ([]byte, error) {
+	return binary.LittleEndian.AppendUint64([]byte{}, c.Instance), nil
+}
+
+var ErrCertAlreadyExists = errors.New("certificate already exists in the CertStore")
+var ErrCertNotFound = errors.New("certificate not found")
+
+// CertStore is responsible for storying and relaying information about new finality certificates
+type CertStore struct {
+	dsLk     sync.Mutex
+	ds       datastore.Datastore
+	busCerts broadcast.Channel[Cert]
+}
+
+func NewCertStore(ctx context.Context, ds datastore.Datastore) (*CertStore, error) {
+	cs := &CertStore{
+		ds: namespace.Wrap(ds, datastore.NewKey("/certstore")),
+	}
+	latestCert, err := cs.loadLatest(ctx)
+	if err != nil {
+		return nil, xerrors.Errorf("loading latest cert: %w", err)
+	}
+	if latestCert != nil {
+		cs.busCerts.Publish(*latestCert)
```

Comment:
How could anyone possibly have subscribed to this yet? 

I suggest moving the side-effects out of the constructor into a new method (that can be called after other things have been constructed and hooked up so they might subscribe here).

Reply:
the broadcast maintains last published entry, this primes it